src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -685,369 +685,243 @@ private static bool TryCopyTo(string source, Span<char> destination, out int cha
-                return Int32ToHexStr((int)value, (char)(fmt - ('X' - 'A' + 10)), digits);
+                // mask will always be trivial
+                Debug.Assert(mask == ~0L);
+                return value;
             }
-            else
-            {
-                NumberFormatInfo info = NumberFormatInfo.GetInstance(provider);
+        }
 
-                byte* pDigits = stackalloc byte[UInt32NumberBufferLength];
-                NumberBuffer number = new NumberBuffer(NumberBufferKind.Integer, pDigits, UInt32NumberBufferLength);
+        private struct UInt64Fmt : IIntFmt<ulong>
+        {
+            public string ToDecStr(ulong value, int digits, IFormatProvider? provider)
+                => UInt64ToDecStr(value, digits);
+            public bool TryToDecStr(ulong value, int digits, IFormatProvider? provider, Span<char> destination, out int charsWritten)
+                => TryUInt64ToDecStr(value, digits, destination, out charsWritten);
 
-                UInt32ToNumber(value, ref number);
+            public string ToHexStr(ulong value, char hexBase, int digits)
+                => Int64ToHexStr((long)value, hexBase, digits);
+            public bool TryToHexStr(ulong value, char hexBase, int digits, Span<char> destination, out int charsWritten)
+                => TryInt64ToHexStr((long)value, hexBase, digits, destination, out charsWritten);
 
-                char* stackPtr = stackalloc char[CharStackBufferSize];
-                ValueStringBuilder sb = new ValueStringBuilder(new Span<char>(stackPtr, CharStackBufferSize));
+            public void ToNumber(ulong input, ref NumberBuffer number)
+                => UInt64ToNumber(input, ref number);
 
-                if (fmt != 0)
-                {
-                    NumberToString(ref sb, ref number, fmt, digits, info);
-                }
-                else
-                {
-                    NumberToStringFormat(ref sb, ref number, format, info);
-                }
-                return sb.ToString();
+            public ulong HexMask(ulong value, ulong mask)
+            {
+                // mask will always be trivial
+                Debug.Assert(mask == ~0UL);
+                return value;
             }
         }
 
-        public static unsafe bool TryFormatUInt32(uint value, ReadOnlySpan<char> format, IFormatProvider? provider, Span<char> destination, out int charsWritten)
+        private static char GetHexBase(char fmt)
+        {
+            // The fmt-(X-A+10) hack has the effect of dictating whether we produce uppercase or lowercase
+            // hex numbers for a-f. 'X' as the fmt code produces uppercase. 'x' as the format code produces lowercase.
+            return (char)(fmt - ('X' - 'A' + 10));
+        }
+
+        private static string FormatInt<TInt, TIntFmt>(TInt value, TInt hexMask, ReadOnlySpan<char> format, IFormatProvider? provider)
```

Comment:
```suggestion
        private static string FormatInt<TInt, TIntFmt>(TInt value, TInt hexMask, string format, IFormatProvider? provider)
```
Is the `format` argument of methods that return `string` ever a `ReadOnlySpan`? If not, it would be better to keep it as `string` for as long as possible.

Reply:
I've incorporated this change. Thanks for spotting it.